Wikipedia/Code/PageHistoryViewController.swift

```diff
@@ -0,0 +1,755 @@
+        view.addSubview(comparisonSelectionViewController.view)
+        comparisonSelectionViewController.didMove(toParent: self)
+        comparisonSelectionViewVisibleConstraint = view.bottomAnchor.constraint(equalTo: comparisonSelectionViewController.view.bottomAnchor)
+        comparisonSelectionViewHiddenConstraint = view.bottomAnchor.constraint(equalTo: comparisonSelectionViewController.view.topAnchor)
+
+        let leadingConstraint = view.leadingAnchor.constraint(equalTo: comparisonSelectionViewController.view.leadingAnchor)
+        let trailingConstraint = view.trailingAnchor.constraint(equalTo: comparisonSelectionViewController.view.trailingAnchor)
+
+        NSLayoutConstraint.activate([comparisonSelectionViewHiddenConstraint!, leadingConstraint, trailingConstraint])
+    }
+
+    private func setComparisonSelectionViewHidden(_ hidden: Bool, animated: Bool) {
+        let changes = {
+            if hidden {
+                self.comparisonSelectionViewVisibleConstraint?.isActive = false
+                self.comparisonSelectionViewHiddenConstraint?.isActive = true
+            } else {
+                self.comparisonSelectionViewHiddenConstraint?.isActive = false
+                self.comparisonSelectionViewVisibleConstraint?.isActive = true
+            }
+            self.view.layoutIfNeeded()
+        }
+        if animated {
+            UIView.animate(withDuration: 0.3, animations: changes)
+        } else {
+            changes()
+        }
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        hintController = PageHistoryHintController()
+        title = CommonStrings.historyTabTitle
+        navigationItem.rightBarButtonItem = compareButton
+        addChild(countsViewController)
+        navigationBar.addUnderNavigationBarView(countsViewController.view)
+        navigationBar.shadowColorKeyPath = \Theme.colors.border
+        countsViewController.didMove(toParent: self)
+
+        navigationBar.isUnderBarViewHidingEnabled = true
+
+        layoutManager.register(PageHistoryCollectionViewCell.self, forCellWithReuseIdentifier: PageHistoryCollectionViewCell.identifier, addPlaceholder: true)
+        collectionView.dataSource = self
+        view.wmf_addSubviewWithConstraintsToEdges(collectionView)
+
+        setupComparisonSelectionViewController()
+
+        apply(theme: theme)
+
+        pageHistoryFetcher.fetchPageCreationDate(for: pageTitle, pageURL: pageURL) { result in
+            switch result {
+            case .failure(let error):
+                self.showNoInternetConnectionAlertOrOtherWarning(from: error)
+            case .success(let firstEditDate):
+                self.pageHistoryFetcher.fetchEditCounts(.edits, for: self.pageTitle, pageURL: self.pageURL) { result in
+                    switch result {
+                    case .failure(let error):
+                        self.showNoInternetConnectionAlertOrOtherWarning(from: error)
```

Comment:
yes, good catch! I need to rethink failures in fetchEditCounts cause there are multiple requests

Reply:
OK, so I return with a failure if I don't get any of the counts and if I managed to catch at least 1 error - https://github.com/wikimedia/wikipedia-ios/pull/3295/commits/0e5ec43501ff1186b31c2d1419600b00d0e7fbb8

In the UI, I'm assuming that I don't want to show an error if I can get at least one count.
In the fetcher, I end up relying on the most recent error, which might not be correct (in theory, calls can fail with different errors). I could return a dictionary of errors grouped by count types or I could split the requests and have the "client" VC match them all together, but even then, I would still want to show only 1 error, and only if I didn't get any counts. 😬

Let me know if you have a different idea

Reply:
You could maybe return a more general error instead of the specific error from that endpoint to avoid the overhead of keeping track of them all - would RequestError.unexpectedResponse work as general enough? Or RequestError.unknown and I think the CommonStrings.genericErrorDescription would get displayed.